lib/matplotlib/figure.py

```diff
@@ -1808,119 +1808,64 @@ def _normalize_grid_string(layout):
             layout = inspect.cleandoc(layout)
             return [list(ln) for ln in layout.strip('\n').split('\n')]
 
-    def subplot_mosaic(self, mosaic, *, sharex=False, sharey=False,
+    def _sub_prep(self, mosaic, sub_add_func, *, sharex=False, sharey=False,
                        width_ratios=None, height_ratios=None,
                        empty_sentinel='.',
-                       subplot_kw=None, per_subplot_kw=None, gridspec_kw=None):
+                       subthing_kw=None, per_subthing_kw=None, gridspec_kw=None):
         """
-        Build a layout of Axes based on ASCII art or nested lists.
-
-        This is a helper function to build complex GridSpec layouts visually.
-
-        See :ref:`mosaic`
-        for an example and full API documentation
+        Helper function that implements the internal logic for `.Figure.subplot_mosaic`
+        and `.Figure.subfigure_mosaic`.
 
         Parameters
         ----------
         mosaic : list of list of {hashable or nested} or str
+            A visual layout of how the user wants their mosaic to be arranged.
 
-            A visual layout of how you want your Axes to be arranged
-            labeled as strings.  For example ::
+        sub_add_func : `.Figure.add_subplot` for `.Figure.subplot_mosaic` and
```

Comment:
ok, so I think the fix to this nested figure problem is to change the signature `sub_add_func`.  The input should stay the same, but the return should be the returned thing _and_ the next `sub_add_func` to use.


For axes it would be

```
def _add_func(*args, **kwargs):
    return self.add_subplot(*args, **kwargs), _add_func

return self._sub_prec(mosaic, _add_func, ...)
```

and for subfigures it would be a bit fancier like

```python
def _factory(fig):
    def _add_sub_func(*arg, **kwargs):
        new_fig = fig.add_subfigure(*args, **kwargs)
        return new_fig, _factory(new_fig)

return self._sub_prec(mosaic, _factory(self), ...)
```


and then where you call it it would be

```
new_thing, sub_add_func = sub_add_func(....)
```

That way we do not have to special case anything in the `_sub_prep` code or have the awkward sometimes we use one input sometimes we use the other behavior on the internal API.

Reply:
I also have not tried this code so it is probably only 90% correct, but I am pretty confident it is pointing in the right direction.

Reply:
Just to not leave this unanswered after my previous comments elsewhere - this will definitely require some thought. Sorry I can't do it right away, but I'll get on it in the morning and, hopefully, make it work :)

Reply:
"recursion" is called that because you curse it again and again.....